File: jdroid-android-core/src/main/java/com/jdroid/android/google/GooglePlayServicesUtils.java
package com.jdroid.android.google;

import android.app.Activity;
import android.app.Dialog;
import android.content.Context;
import androidx.annotation.WorkerThread;

import com.google.android.gms.ads.identifier.AdvertisingIdClient;
import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.GoogleApiAvailability;
import com.jdroid.android.application.AbstractApplication;
import com.jdroid.java.utils.LoggerUtils;

import org.slf4j.Logger;

public class GooglePlayServicesUtils {

	private static final Logger LOGGER = LoggerUtils.getLogger(GooglePlayServicesUtils.class);

	private static final String GOOGLE_PLAY_SERVICES = "com.google.android.gms";

	private static final int GOOGLE_PLAY_SERVICES_RESOLUTION_REQUEST = 1;

	public static GooglePlayServicesResponse verifyGooglePlayServices(Activity activity) {
		GoogleApiAvailability googleApiAvailability = GoogleApiAvailability.getInstance();
		int resultCode = googleApiAvailability.isGooglePlayServicesAvailable(AbstractApplication.get());
		LOGGER.info("Google Play Services result code: " + resultCode);
		if (resultCode == ConnectionResult.SUCCESS) {
			return new GooglePlayServicesResponse();
		} else {
			Dialog dialog = null;
			if (googleApiAvailability.isUserResolvableError(resultCode)) {
				dialog = googleApiAvailability.getErrorDialog(activity, resultCode, GOOGLE_PLAY_SERVICES_RESOLUTION_REQUEST);
				if (dialog != null) {
					// This dialog will help the user update to the latest GooglePlayServices
					dialog.show();
				}
			}
			return new GooglePlayServicesResponse(dialog, resultCode);
		}
	}

	public static void checkGooglePlayServices(Activity activity) {
		GoogleApiAvailability googleApiAvailability = GoogleApiAvailability.getInstance();
		int resultCode = googleApiAvailability.isGooglePlayServicesAvailable(activity);
		LOGGER.info("Google Play Services result code: " + resultCode);
		if (resultCode != ConnectionResult.SUCCESS && googleApiAvailability.isUserResolvableError(resultCode)) {
			googleApiAvailability.showErrorDialogFragment(activity, resultCode, GOOGLE_PLAY_SERVICES_RESOLUTION_REQUEST);
		}
	}

	public static boolean isGooglePlayServicesAvailable(Context context) {
		GoogleApiAvailability googleApiAvailability = GoogleApiAvailability.getInstance();
		int resultCode = googleApiAvailability.isGooglePlayServicesAvailable(context);
		LOGGER.info("Google Play Services result code: " + resultCode);
		return resultCode == ConnectionResult.SUCCESS;
	}

	public static void launchGooglePlayServicesUpdate() {
		GooglePlayUtils.INSTANCE.launchAppDetails(GOOGLE_PLAY_SERVICES);
	}

	@WorkerThread
	public static String getAdvertisingId() {
		try {
			return AdvertisingIdClient.getAdvertisingIdInfo(AbstractApplication.get()).getId();
		} catch (Exception e) {
			LOGGER.warn("Error getting the advertising id " + e.getMessage());
			return null;
		}
	}

	public static class GooglePlayServicesResponse {

		private Boolean isAvailable;
		private Dialog dialog;
		private Integer resultCode;

		public GooglePlayServicesResponse() {
			isAvailable = true;
		}

		public GooglePlayServicesResponse(Dialog dialog, Integer resultCode) {
			isAvailable = false;
			this.dialog = dialog;
			this.resultCode = resultCode;
		}

		public Boolean isAvailable() {
			return isAvailable;
		}

		public Dialog getDialog() {
			return dialog;
		}

		public Integer getResultCode() {
			return resultCode;
		}
	}
}

File: jdroid-android-core/src/main/java/com/jdroid/android/google/GooglePlayServicesUtils.kt
package com.jdroid.android.google

import android.app.Activity
import android.app.Dialog
import android.content.Context
import androidx.annotation.WorkerThread
import com.google.android.gms.ads.identifier.AdvertisingIdClient
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.GoogleApiAvailability
import com.jdroid.android.application.AbstractApplication
import com.jdroid.java.utils.LoggerUtils

object GooglePlayServicesUtils {

    private val LOGGER = LoggerUtils.getLogger(GooglePlayServicesUtils::class.java)

    private const val GOOGLE_PLAY_SERVICES = "com.google.android.gms"
    private const val GOOGLE_PLAY_SERVICES_RESOLUTION_REQUEST = 1

    @WorkerThread
    fun getAdvertisingId(): String? {
        return try {
            AdvertisingIdClient.getAdvertisingIdInfo(AbstractApplication.get()).id
        } catch (e: Exception) {
            LOGGER.warn("Error getting the advertising id " + e.message)
            null
        }
    }

    fun verifyGooglePlayServices(activity: Activity): GooglePlayServicesResponse {
        val googleApiAvailability = GoogleApiAvailability.getInstance()
        val resultCode = googleApiAvailability.isGooglePlayServicesAvailable(AbstractApplication.get())
        LOGGER.info("Google Play Services result code: $resultCode")
        return if (resultCode == ConnectionResult.SUCCESS) {
            GooglePlayServicesResponse()
        } else {
            var dialog: Dialog? = null
            if (googleApiAvailability.isUserResolvableError(resultCode)) {
                dialog = googleApiAvailability.getErrorDialog(activity, resultCode, GOOGLE_PLAY_SERVICES_RESOLUTION_REQUEST)
                dialog?.show()
            }
            GooglePlayServicesResponse(dialog, resultCode)
        }
    }

    fun checkGooglePlayServices(activity: Activity) {
        val googleApiAvailability = GoogleApiAvailability.getInstance()
        val resultCode = googleApiAvailability.isGooglePlayServicesAvailable(activity)
        LOGGER.info("Google Play Services result code: $resultCode")
        if (resultCode != ConnectionResult.SUCCESS && googleApiAvailability.isUserResolvableError(resultCode)) {
            googleApiAvailability.showErrorDialogFragment(activity, resultCode, GOOGLE_PLAY_SERVICES_RESOLUTION_REQUEST)
        }
    }

    fun isGooglePlayServicesAvailable(context: Context): Boolean {
        val googleApiAvailability = GoogleApiAvailability.getInstance()
        val resultCode = googleApiAvailability.isGooglePlayServicesAvailable(context)
        LOGGER.info("Google Play Services result code: $resultCode")
        return resultCode == ConnectionResult.SUCCESS
    }

    fun launchGooglePlayServicesUpdate() {
        GooglePlayUtils.launchAppDetails(GOOGLE_PLAY_SERVICES)
    }

    class GooglePlayServicesResponse {

        var isAvailable: Boolean
        var dialog: Dialog? = null
        var resultCode: Int? = null

        constructor() {
            isAvailable = true
        }

        constructor(dialog: Dialog?, resultCode: Int) {
            isAvailable = false
            this.dialog = dialog
            this.resultCode = resultCode
        }
    }
}
